hazelcast-client/src/test/java/com/hazelcast/client/listeners/AbstractListenersOnReconnectTest.java

```diff
@@ -206,225 +152,267 @@ public void stateChanged(LifecycleEvent event) {
-        clientConfig.getNetworkConfig().setConnectionAttemptLimit(Integer.MAX_VALUE);
+        ClientConfig clientConfig = getSmartClientConfig();
         client = factory.newHazelcastClient(clientConfig);
 
-        testListenersTerminateRandomNodeInternal();
+        testListenersHeartbeatTimeoutToOwner();
     }
 
     @Test
-    public void testListenersSmartRoutingTerminateRandomNode()
-            throws InterruptedException {
+    public void testClusterReconnectDueToHeartbeatNonSmartRouting() {
         factory.newHazelcastInstance();
 
-        ClientConfig clientConfig = createClientConfig();
+        ClientConfig clientConfig = getNonSmartClientConfig();
         client = factory.newHazelcastClient(clientConfig);
-        testListenersTerminateRandomNodeInternal();
+
+        testListenersHeartbeatTimeoutToOwner();
     }
 
     @Test
-    public void testListenersMemberDestroyEndpointTaskSmartRouting()
-            throws InterruptedException {
+    public void testClusterReconnectMultipleServerDueToHeartbeatNonSmartRouting() {
         factory.newInstances(null, 3);
 
-        ClientConfig clientConfig = createClientConfig();
-        clientConfig
-                .setProperty(ClientProperty.HEARTBEAT_TIMEOUT.getName(), String.valueOf(2 * ENDPOINT_REMOVE_DELAY_MILLISECONDS));
+        ClientConfig clientConfig = getNonSmartClientConfig();
         client = factory.newHazelcastClient(clientConfig);
-        testListenersWaitMemberDestroy();
+
+        testListenersHeartbeatTimeoutToOwner();
     }
 
-    @Test
-    public void testClusterReconnectDueToHeartbeatSmartRouting()
-            throws InterruptedException {
-        factory.newHazelcastInstance();
+    private void testListenersHeartbeatTimeoutToOwner() {
+        setupListener();
+
+        HazelcastClientInstanceImpl clientInstanceImpl = getHazelcastClientInstanceImpl(client);
+        HazelcastInstance server = getOwnerServer(factory, clientInstanceImpl);
 
-        ClientConfig clientConfig = getSmartClientConfig();
-        client = factory.newHazelcastClient(clientConfig);
 
-        testListenersForHeartbeat();
+        final CountDownLatch disconnectedLatch = new CountDownLatch(1);
+        final CountDownLatch connectedLatch = new CountDownLatch(1);
+        client.getLifecycleService().addLifecycleListener(new LifecycleListener() {
+            @Override
+            public void stateChanged(LifecycleEvent event) {
+                if (LifecycleEvent.LifecycleState.CLIENT_DISCONNECTED == event.getState()) {
+                    disconnectedLatch.countDown();
+                }
+                if (LifecycleEvent.LifecycleState.CLIENT_CONNECTED == event.getState()) {
+                    connectedLatch.countDown();
+                }
+            }
+        });
+
+        blockMessagesFromInstance(server, client);
+        assertOpenEventually(disconnectedLatch);
+
+        unblockMessagesFromInstance(server, client);
+        assertOpenEventually(connectedLatch);
```

Comment:
Actually we are testing both, since it will connect to a random node. But we can work on two separate tests with `hazelcast.client.shuffle.member.list` is off and two carefully prepared memberlist.